### ISSUE TITLE
Create a crate for EdDSA Poseidon (BabyJubjub curve)

### ISSUE THREAD

**Issue by cedoor** [privacy-scaling-explorations/zk-kit.rust#9]
**Describe the package you'd like**

ZK-Kit already provides a [package](https://github.com/privacy-scaling-explorations/zk-kit/tree/main/packages/eddsa-poseidon) to generate EdDSA keys that uses [Poseidon](https://www.poseidon-hash.info/) hashes and is based on the [BabyJubjub elliptic curve](https://eips.ethereum.org/EIPS/eip-2494). The Rust crate should be compatible with the JS version.

Possible dependencies:
* https://github.com/arnaucube/babyjubjub-rs
* https://github.com/arnaucube/poseidon-rs

It is worth checking whether an implementation in Rust already exists.

**Comment by cedoor** [comment]
Hey @arnaucube, may I know if you're still maintaining `babyjubjub-rs` and `poseidon-rs`? 

Might you be interested in including them in the ZK-Kit package set? You would remain the author ofc.

For context. ZK-Kit is a collective effort of PSE devs that aims to provide a set of reusable libraries in a context where standards on code quality, documentation, test coverage, and audits are guaranteed.

**Comment by arnaucube** [comment]
Hi, sure, they both are compatible with [circomlib](https://github.com/iden3/circomlib)'s version (also with [circomlibjs](https://github.com/iden3/circomlibjs)'s version).
And if useful, there exists also compatible implementations of both Poseidon & BabyJubJub's EdDSA in Go in https://github.com/iden3/go-iden3-crypto 's repo.

If useful, there is also other  a merkletree implementation in Go that is compatible with circomlib's sparsemerkletree version, with the main feature (appart from full compatibility) that it parallelizes by CPUs so it goes faster than the js one (eg. 2m09s vs 0.436s adding 10k leafs): https://github.com/vocdoni/arbo .

Regarding being maintained, the mentioned repos are not actively being developed since they reached the desired features, but if there are future fixes to potential bugs they will be taken care of.

**Comment by kilic** [comment]
Another way might be just using arkworks backend as in here https://github.com/kilic/arkeddsa/ both for poseidon and jubjub

**Comment by cedoor** [comment]
@kilic interesting. Is Poseidon compatible with the Iden3 version?

**Comment by kilic** [comment]
> @kilic interesting. Is Poseidon compatible with the Iden3 version?

Not as it is in kilic/arkeddsa but so close

1. We should give the matching config params
2. Circom takes 0th element of the state as output and arkworks skips 0th element. However in arkworks [state](https://github.com/arkworks-rs/crypto-primitives/blob/6b195553444650fb37663ee1919c9fd885f5dbd9/crypto-primitives/src/sponge/poseidon/mod.rs#L54) is public so 0th element is accessible and then it becomes circom compatible

**Comment by arnaucube** [comment]
oh right! Agree with @kilic ^^, if you can do with with arkworks it would be much better than depending on the initially mentioned libs, less dependencies to maintain, and more compatible with other projects (in the arkworks ecosystem).

**Comment by cedoor** [comment]
Thank you guys 🙏🏽 I'll let this issue open for anyone who wants to implement it and follow your suggestions.

**Comment by 1010adigupta** [comment]
@cedoor I would like to take this up

**Comment by cedoor** [comment]
@1010adigupta Sure, I'll assign this issue to you then :)

**Comment by sripwoud** [comment]
@1010adigupta Are you still working on this? (If not we may assign it to someone else)

**Comment by 1010adigupta** [comment]
> @1010adigupta Are you still working on this? (If not we may assign it to someone else)

yes, you may assign it to someone else, got busy with some other stuff

**Comment by ozgurarmanc** [comment]
@cedoor @sripwoud  Can I get this issue?

**Comment by cedoor** [comment]
Hey @ozgurarmanc, ofc 👍🏽

**Comment by Some-of-the-things** [comment]
I can take a crack at picking this one up if you'd like...

**Comment by cedoor** [comment]
@Some-of-the-things ofc, I'll assign it to you
